File: testcases/src/main/java/BrowsingAndCategorization.java
import java.util.LinkedHashMap;
import java.util.Map;

class Product {
    String name;
    boolean inStock;
    String category;

    Product(String name, boolean inStock, String category) {
        this.name = name;
        this.inStock = inStock;
        this.category = category;
    }
}

public class BrowsingAndCategorization {
    private Map<String, Product> productCatalog;

    public BrowsingAndCategorization() {
        productCatalog = new LinkedHashMap<>(); // preserves insertion order
        productCatalog.put("Keyboard", new Product("Keyboard", true, "Technology"));
        productCatalog.put("Mouse", new Product("Mouse", false, "Technology"));
        productCatalog.put("Shoes", new Product("Shoes", true, "Fashion"));
    }

    public String clickProduct(String productName) {
        Product product = productCatalog.get(productName);
        if (product == null) return "Product not found";
        if (!product.inStock)
            return product.name + " - Sorry this item is out of stock.";
        return product.name + " - Product Information Displayed";
    }

    public String clickCategory(String category) {
        StringBuilder sb = new StringBuilder("Products in " + category + ": ");
        for (Product product : productCatalog.values()) {
            if (product.category.equals(category)) {
                sb.append(product.name).append(" ");
            }
        }
        return sb.toString().trim();
    }

    public String enterWebsite() {
        return "Product browsing page displayed";
    }
}

File: server/BrowsingAndCategorization.js
class Product {
  constructor(name, inStock, category) {
    this.name = name;
    this.inStock = inStock;
    this.category = category;
  }
}

class BrowsingAndCategorization {
  constructor() {
    this.productCatalog = new Map();
    this.productCatalog.set("Keyboard", new Product("Keyboard", true, "Technology"));
    this.productCatalog.set("Mouse", new Product("Mouse", false, "Technology"));
    this.productCatalog.set("Shoes", new Product("Shoes", true, "Fashion"));
  }

  clickProduct(productName) {
    const product = this.productCatalog.get(productName);
    if (!product) return "Product not found";
    if (!product.inStock) return `${product.name} - Sorry this item is out of stock.`;
    return `${product.name} - Product Information Displayed`;
  }

  clickCategory(category) {
    const products = [];
    for (const product of this.productCatalog.values()) {
      if (product.category === category) {
        products.push(product.name);
      }
    }
    return `Products in ${category}: ${products.join(" ")}`;
  }

  enterWebsite() {
    return "Product browsing page displayed";
  }
}

module.exports = BrowsingAndCategorization;
